// @sanity/orderable-document-list/lib/Document.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Document;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _preview = _interopRequireDefault(require("part:@sanity/base/preview"));

var _schema = _interopRequireDefault(require("part:@sanity/base/schema"));

var _OrderableContext = require("./OrderableContext");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// eslint-disable-next-line no-unused-vars
function Document(_ref) {
  var doc = _ref.doc,
      increment = _ref.increment,
      entities = _ref.entities,
      handleSelect = _ref.handleSelect,
      index = _ref.index,
      isFirst = _ref.isFirst,
      isLast = _ref.isLast;

  var _useContext = (0, _react.useContext)(_OrderableContext.OrderableContext),
      showIncrements = _useContext.showIncrements;

  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingX: 3,
    style: {
      flexShrink: 0
    }
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    fontSize: 4
  }, /*#__PURE__*/_react.default.createElement(_icons.DragHandleIcon, null))), showIncrements && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    style: {
      flexShrink: 0
    },
    align: "center",
    gap: 1,
    paddingRight: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    padding: 2,
    mode: "ghost",
    onClick: () => increment(index, index + -1, doc._id, entities),
    disabled: isFirst,
    icon: _icons.ChevronUpIcon
  }), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    padding: 2,
    mode: "ghost",
    disabled: isLast,
    onClick: () => increment(index, index + 1, doc._id, entities),
    icon: _icons.ChevronDownIcon
  })), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    style: {
      width: "100%"
    },
    padding: 2,
    mode: "bleed",
    onClick: e => handleSelect(doc._id, index, e.nativeEvent)
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    flex: 1,
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_preview.default, {
    value: doc,
    type: _schema.default.get(doc._type)
  })))));
}

Document.propTypes = {
  doc: _propTypes.default.shape({
    _id: _propTypes.default.string,
    _type: _propTypes.default.string
  }).isRequired,
  entities: _propTypes.default.arrayOf(_propTypes.default.shape({
    _id: _propTypes.default.string
  }).isRequired).isRequired,
  handleSelect: _propTypes.default.func.isRequired,
  increment: _propTypes.default.func.isRequired,
  index: _propTypes.default.number.isRequired,
  isFirst: _propTypes.default.bool.isRequired,
  isLast: _propTypes.default.bool.isRequired
};
