// @sanity/orderable-document-list/lib/OrderableDocumentList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _DocumentListWrapper = _interopRequireDefault(require("./DocumentListWrapper"));

var _resetOrder3 = require("./helpers/resetOrder");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Must use a Class Component here so the actionHandlers can be called
class OrderableDocumentList extends _react.Component {
  constructor(props) {
    var _this;

    super(props);
    _this = this;

    _defineProperty(this, "actionHandlers", {
      showIncrements: () => {
        this.setState(state => ({
          showIncrements: !state.showIncrements
        }));
      },
      resetOrder: function () {
        var _resetOrder2 = _asyncToGenerator(function* () {
          var _update$results;

          _this.setState(() => ({
            resetOrderTransaction: {
              status: "info",
              title: "Reordering started...",
              closable: true
            }
          }));

          var update = yield (0, _resetOrder3.resetOrder)(_this.props.options.type);
          var reorderWasSuccessful = update === null || update === void 0 ? void 0 : (_update$results = update.results) === null || _update$results === void 0 ? void 0 : _update$results.length;

          _this.setState(() => ({
            resetOrderTransaction: {
              status: reorderWasSuccessful ? "success" : "info",
              title: reorderWasSuccessful ? "Reordered ".concat(update.results.length === 1 ? "Document" : "Documents") : "Reordering failed",
              closable: true
            }
          }));
        });

        function resetOrder() {
          return _resetOrder2.apply(this, arguments);
        }

        return resetOrder;
      }()
    });

    this.state = {
      showIncrements: false,
      resetOrderTransaction: {}
    };
  }

  render() {
    var _this$props, _this$props$options, _this$props2, _this$props2$options, _this$props3, _this$props3$options;

    return /*#__PURE__*/_react.default.createElement(_DocumentListWrapper.default, {
      type: this === null || this === void 0 ? void 0 : (_this$props = this.props) === null || _this$props === void 0 ? void 0 : (_this$props$options = _this$props.options) === null || _this$props$options === void 0 ? void 0 : _this$props$options.type,
      filter: this === null || this === void 0 ? void 0 : (_this$props2 = this.props) === null || _this$props2 === void 0 ? void 0 : (_this$props2$options = _this$props2.options) === null || _this$props2$options === void 0 ? void 0 : _this$props2$options.filter,
      params: this === null || this === void 0 ? void 0 : (_this$props3 = this.props) === null || _this$props3 === void 0 ? void 0 : (_this$props3$options = _this$props3.options) === null || _this$props3$options === void 0 ? void 0 : _this$props3$options.params,
      showIncrements: this.state.showIncrements,
      resetOrderTransaction: this.state.resetOrderTransaction
    });
  }

}

exports.default = OrderableDocumentList;

_defineProperty(OrderableDocumentList, "propTypes", {
  options: _propTypes.default.shape({
    type: _propTypes.default.string,
    filter: _propTypes.default.string,
    params: _propTypes.default.object
  }).isRequired
});
